File: zaiko/zaiko/src/utils/type_check.zig
// Type checking utilities matching C implementation
// Based on maiko/inc/lsptypes.h

const types = @import("types.zig");
const stack = @import("../vm/stack.zig");

const LispPTR = types.LispPTR;
const DLword = types.DLword;
const VM = stack.VM;

// Type numbers from maiko/inc/lsptypes.h
pub const TYPE_SMALLP: u16 = 1; // Small integer (encoded directly in LispPTR)
pub const TYPE_FIXP: u16 = 2; // Boxed integer (pointer to memory)
pub const TYPE_LISTP: u16 = 5;
pub const TYPE_NEWATOM: u16 = 21;

// Special values
pub const ATOM_T: LispPTR = 1; // T atom
pub const NIL_PTR: LispPTR = 0; // NIL

/// Get type entry from type table
/// C: GetTypeEntry(address) = GETWORD(MDStypetbl + ((address) >> 9))
/// For now, simplified version - will need full type table access later
fn getTypeEntry(vm: *VM, address: LispPTR) ?u16 {
    _ = vm; // Will use VM to access type table when available
    
    // Special cases
    if (address == NIL_PTR) return 0; // NIL type
    if (address == ATOM_T) return TYPE_NEWATOM; // T is an atom
    
    // For now, we can't fully implement without type table access
    // This is a placeholder that will be extended
    return null;
}

/// Get type number from address
/// C: GetTypeNumber(address) = GetTypeEntry(address) & 0x7ff
pub fn getTypeNumber(vm: *VM, address: LispPTR) ?u16 {
    const type_entry = getTypeEntry(vm, address) orelse return null;
    return type_entry & 0x7ff; // Low 11 bits
}

/// Check if address is a list (cons cell)
/// C: Listp(address) = (GetTypeNumber(address) == TYPE_LISTP)
pub fn isList(vm: *VM, address: LispPTR) bool {
    // Special cases
    if (address == NIL_PTR) return false; // NIL is not a list (though (cdr nil) = nil)
    if (address == ATOM_T) return false; // T is an atom, not a list
    
    // For addresses that could be cons cells:
    // - Must be even (4-byte aligned)
    // - Must be within virtual memory bounds
    if ((address & 1) != 0) return false; // Odd addresses are fixnums
    
    // Check if we can access the address (basic validation)
    if (vm.virtual_memory) |vmem| {
        if (@as(usize, @intCast(address)) >= vmem.len) return false;
    } else {
        return false;
    }
    
    // For now, assume even addresses within bounds could be cons cells
    // Full implementation would check type table: GetTypeNumber(address) == TYPE_LISTP
    // TODO: Implement full type table lookup when MDStypetbl is available
    return true;
}

/// Check if address is NIL
pub fn isNil(address: LispPTR) bool {
    return address == NIL_PTR;
}

/// Check if address is T
pub fn isT(address: LispPTR) bool {
    return address == ATOM_T;
}

File: taiko/src/vm/memory/typecheck.ts
// Type checking utilities for Lisp values
// Per maiko/inc/lsptypes.h and maiko/src/typetbl.c

import type { LispPTR } from '../../utils/types';
import { S_POSITIVE, S_NEGATIVE, NIL_PTR, POINTERMASK } from '../../utils/constants';
import { MemoryManager } from './manager';

/**
 * Type number constants
 * Per maiko/inc/lsptypes.h:61-66
 */
export const TYPE_SMALLP = 1;     // Small integer (immediate)
export const TYPE_FIXP = 2;       // Fixnum (bignum)
export const TYPE_FLOATP = 3;     // Floating point number
export const TYPE_LISTP = 5;      // List (cons cell)

/**
 * Check if value is an immediate (SMALLP, character, etc.)
 * Per C: IF_IMMEDIATE macro
 *
 * Immediate values have high bits set (S_POSITIVE, S_NEGATIVE, etc.)
 */
export function isImmediate(value: LispPTR): boolean {
    // SMALLP positive: top 3 bits set (S_POSITIVE)
    // SMALLP negative: top 2 bits set (S_NEGATIVE)
    // Characters and other immediates also have high bits set
    return (value & 0xC0000000) !== 0 && (value & POINTERMASK) === value;
}

/**
 * Get type number of a Lisp value
 * Per C: GetTypeNumber in maiko/src/typetbl.c
 *
 * Simplified version - for now, only handles:
 * - SMALLP (immediate integers)
 * - FIXP (bignum fixnums)
 * - FLOATP (floating point)
 * - LISTP (cons cells)
 *
 * TODO: Full implementation would check memory structures
 */
export function getTypeNumber(value: LispPTR, vm: { virtualMemory: Uint8Array | null } | null = null): number {
    // NIL is a special case
    if (value === NIL_PTR) {
        return TYPE_LISTP; // NIL is a list
    }

    // Check for SMALLP (immediate integers)
    if ((value & S_POSITIVE) === S_POSITIVE) {
        return TYPE_SMALLP;
    }
    if ((value & S_NEGATIVE) === S_NEGATIVE) {
        return TYPE_SMALLP;
    }

    // For non-immediate values, we'd need to check memory
    // For now, assume they're pointers to typed objects
    // This is a simplified implementation
    // TODO: Read type from memory structure

    // If it's a valid pointer (within POINTERMASK), it could be:
    // - FIXP (bignum)
    // - FLOATP
    // - LISTP (cons cell)
    // - Other types

    // For now, return a default type
    // Full implementation would read the type from memory
    return TYPE_LISTP; // Default to list for now
}

/**
 * Check if value is a number (SMALLP, FIXP, or FLOATP)
 * Per C: Numberp macro
 */
export function isNumber(value: LispPTR, vm: { virtualMemory: Uint8Array | null } | null = null): boolean {
    const type = getTypeNumber(value, vm);
    return type === TYPE_SMALLP || type === TYPE_FIXP || type === TYPE_FLOATP;
}

/**
 * Get FIXP value from memory
 * Per C: FIXP_VALUE macro
 *
 * FIXP values are stored in memory as 32-bit integers
 */
export function getFIXPValue(value: LispPTR, vm: { virtualMemory: Uint8Array | null }): number | null {
    if (!vm.virtualMemory) {
        return null;
    }

    const offset = MemoryManager.Address.lispPtrToByte(value);
    if (offset + 4 > vm.virtualMemory.length) {
        return null;
    }

    const view = new DataView(vm.virtualMemory.buffer, vm.virtualMemory.byteOffset + offset);
    return view.getInt32(0, true); // little-endian
}

/**
 * Get FLOATP value from memory
 * Per C: FLOATP_VALUE macro
 *
 * FLOATP values are stored in memory as 64-bit floats
 */
export function getFLOATPValue(value: LispPTR, vm: { virtualMemory: Uint8Array | null }): number | null {
    if (!vm.virtualMemory) {
        return null;
    }

    const offset = MemoryManager.Address.lispPtrToByte(value);
    if (offset + 8 > vm.virtualMemory.length) {
        return null;
    }

    const view = new DataView(vm.virtualMemory.buffer, vm.virtualMemory.byteOffset + offset);
    return view.getFloat64(0, true); // little-endian
}
